### ISSUE TITLE
[转载]X16还是X8？PCI-E通道数量对机器学习的影响（4 x Titan V测试） - 知乎

### ISSUE THREAD

**Issue by madobet** [madobet/wwwclip#104]
本文转自 [http://www.pugetsystems.com](https://link.zhihu.com/?target=http%3A//www.pugetsystems.com)

![](https://pic2.zhimg.com/v2-a0aebcdbfd0c94707f65a67d6bc5f519_b.jpg)

## **前言**

在使用 GPU 来加速运算时有一个问题一直被激烈讨论：用 PCI-E X16 和 X8 到底对有没有性能差异？我们曾经在许多软件中讨论这个问题，答案通常是：有，但不大。

这个问题被提出了有些日子，在之前我的一个回答中也简单提过它（[深度学习用途的 PC，GPU PCI-E 通道走 8X 会影响性能吗？](https://www.zhihu.com/question/66926337/answer/247696350)）在这之后做了更多关于机器学习的测试，得到了本文中的结论。

之前我们在运行机器学习代码时，无论是用 CPU 和芯片组自带的通道还是用了 PLX（PEX）PCI-E 交换机，我们的 GPU 全部都跑在 X16 通道上。但当一块主板支持 4 路显卡但是又没有配备 PCI-E 交换机时，所有的显卡都跑在 X8 通道上了。这时会有多少性能下降？后面的文章会告诉你答案。

本文会分别使用 PCI-E X16 和 X8 速度在一台配备 4 块 Titan V 显卡的机器上运行一些流行的机器学习（深度学习）代码。

## **测试平台**

**硬件**

-   带 4 X16 PCIe 插槽的技嘉主板 (插槽 2、3 使用 1 PLX 交换机)
-   Intel Xeon W-2195 18 core (Skylake-W with AVX512)
-   256GB Reg ECC 内存 (最高 512GB)
-   4 x NVIDIA Titan V GPU's
-   Samsung 256GB NVMe M.2

**软件**

-   Ubuntu 16.04
-   Docker 18.03.0-ce
-   NVIDIA Docker V2
-   TensorFlow 1.7 (running on NVIDIA NGC docker image)
-   Keras 2.1.5 (with TensorFlow back-end from local install with Anaconda Python).

## **测试方法**

为了让所有的显卡都运行在 X8 通道下，我们改造了一下显卡——用纸覆盖一半金手指。题图就是改造后的四块 Titan V 显卡。不建议用户在家尝试这个方法。

## **测试内容**

GoogLeNet Training：这个代码运行在[NVIDIA NGC](https://link.zhihu.com/?target=https%3A//www.nvidia.com/en-us/gpu-cloud/)上的 TensorFlow docker 镜像里。应用保存在 nvidia-examples 目录下的 cnn 里。测试使用人工合成的输入数据。详细信息可以查看之前的文章：[\[转载\]NVIDIA Titan V 深度学习性能测试（1080Ti/Titan V/Titan V + Tensor cores 对比测试）](https://zhuanlan.zhihu.com/p/35997023)

Billion Words Benchmark LSTM Train：这个代码也在上面那个 TensorFlow docker 镜像里。应用名是 big_lstm，使用十亿字的新闻馈送语料库。

VGG model in Keras：这是个真实案例，使用 Keras（使用 TensorFlow 后端）实现 VGG CNN。在一台安装了 Anaconda Python 的 Jupyter notebook 中运行。数据是 25000 张图像（来源：["dogs vs cats" downloaded from Kaggle](https://link.zhihu.com/?target=https%3A//www.kaggle.com/c/dogs-vs-cats/data)）

p2p 带宽和延迟：使用了 NVIDIA CUDA 案例中的 p2pBandwidthLatencyTest 来展示 X16 和 X8 通道下各显卡点到点通讯的带宽。

## **结果**

-   GoogLeNet Training, TensorFlow FP32 and FP16 (Tensor-cores)

\[单位：张图像 / 秒 (total batch size)]

![](https://pic1.zhimg.com/v2-41cbed5dcc381a72f6833e5137a2d324_b.jpg)
![](https://pic4.zhimg.com/v2-34ebb6d66101956a8ff08c2810974d7f_b.jpg)

数据显示代码 GPU 优化非常好，且 X16 和 X8 只有很小的差异。在增加显卡数量后，差异放大。数据同样展示了使用 Tensor-cores (FP16) 后的巨大性能提升。

-   Billion Words Benchmark LSTM Train

\[单位：单词 / 秒]

![](https://pic4.zhimg.com/v2-b25ae08844091c9e3d18016a7d75b3b7_b.jpg)
![](https://pic1.zhimg.com/v2-54b8a6b901e9be7ce53d31b0d1bddcf0_b.jpg)

-   VGG, Keras (TensorFlow) Memory-streaming 25000 Images

\[单位：四次迭代耗时（秒）]

![](https://pic2.zhimg.com/v2-6d5573bb49b60649d0ca7131c807587d_b.jpg)
![](https://pic3.zhimg.com/v2-24859b0b689122244916748eaee23a36_b.jpg)

这个测试中，X16 和 X8 之间的性能差异更小。 超过 2 块 GPU 后，多卡效率也不如之前两个项目。 图像数据被加载到内存中，并通过 Python 变量提供给模型。 文章最后有这个项目的 Python 源代码。

下面的测试数据提供了一个明确的结论：X16 相比 X8 只有很小的优势，在一些情况下甚至没有明显的性能差异。但这不意味着所有情况都是这样，我们测试的代码都是通过将任务分配给各 GPU 来实现并行化的，因此卡和卡之间的通讯非常少。这种情况下带宽对性能的影响局限于 CPU 和 GPU 之间传递数据的时候。带宽对性能的影响取决于具体的工作内容，但在大多数情况下影响不大。

继续降低 PCI-E 通道数量会发生什么？X4 会怎么样？如果用 X1 的 “采矿设备” 来机器学习又会怎样？如果你这样做了，欢迎在文章下留言。

## **附录 A：p2p 带宽和延迟测试结果**

PCIe X16

    ./p2pBandwidthLatencyTest

    [P2P (Peer-to-Peer) GPU Bandwidth Latency Test]
    Device: 0, TITAN V, pciBusID: 17, pciDeviceID: 0, pciDomainID:0
    Device: 1, TITAN V, pciBusID: 65, pciDeviceID: 0, pciDomainID:0
    Device: 2, TITAN V, pciBusID: b5, pciDeviceID: 0, pciDomainID:0
    Device: 3, TITAN V, pciBusID: b6, pciDeviceID: 0, pciDomainID:0
    Device=0 CAN Access Peer Device=1
    Device=0 CAN Access Peer Device=2
    Device=0 CAN Access Peer Device=3
    Device=1 CAN Access Peer Device=0
    Device=1 CAN Access Peer Device=2
    Device=1 CAN Access Peer Device=3
    Device=2 CAN Access Peer Device=0
    Device=2 CAN Access Peer Device=1
    Device=2 CAN Access Peer Device=3
    Device=3 CAN Access Peer Device=0
    Device=3 CAN Access Peer Device=1
    Device=3 CAN Access Peer Device=2

    ***NOTE: In case a device doesn't have P2P access to other one,
     it falls back to normal memcopy procedure.
    So you can see lesser Bandwidth (GB/s) in those cases.

    P2P Connectivity Matrix
         D\D     0     1     2     3
         0	     1     1     1     1
         1	     1     1     1     1
         2	     1     1     1     1
         3	     1     1     1     1
    Unidirectional P2P=Disabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 555.65   5.75   5.74   5.76
         1   5.86 554.87   5.72   5.74
         2   5.87   5.87 554.87   5.77
         3   5.75   5.76   5.81 555.65
    Unidirectional P2P=Enabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 554.87   6.03   6.02   6.02
         1   6.05 554.87   6.01   6.03
         2   4.39   4.39 554.08   4.26
         3   4.40   4.40   4.27 553.29
    Bidirectional P2P=Disabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 566.53  11.05  11.03  10.99
         1  11.07 564.49  11.15  10.92
         2  11.15  11.16 562.86   6.18
         3  11.05  11.01   6.19 564.49
    Bidirectional P2P=Enabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 564.05  10.83   8.40   8.43
         1  10.84 563.67   8.40   8.43
         2   8.40   8.39 564.90   8.07
         3   8.43   8.42   8.11 563.67
    P2P=Disabled Latency Matrix (us)
       D\D     0      1      2      3
         0   2.94  16.23  16.50  16.51
         1  16.41   3.03  16.53  16.52
         2  17.40  17.54   3.74  18.90
         3  17.42  17.36  18.73   3.08
    P2P=Enabled Latency Matrix (us)
       D\D     0      1      2      3
         0   3.01   5.38   5.55   5.57
         1   5.26   3.00   5.79   5.80
         2   6.81   6.91   3.00   6.64
         3   6.81   6.89   6.62   3.01

PCIe X8

    ./p2pBandwidthLatencyTest
    [P2P (Peer-to-Peer) GPU Bandwidth Latency Test]
    Device: 0, TITAN V, pciBusID: 17, pciDeviceID: 0, pciDomainID:0
    Device: 1, TITAN V, pciBusID: 65, pciDeviceID: 0, pciDomainID:0
    Device: 2, TITAN V, pciBusID: b5, pciDeviceID: 0, pciDomainID:0
    Device: 3, TITAN V, pciBusID: b6, pciDeviceID: 0, pciDomainID:0
    Device=0 CAN Access Peer Device=1
    Device=0 CAN Access Peer Device=2
    Device=0 CAN Access Peer Device=3
    Device=1 CAN Access Peer Device=0
    Device=1 CAN Access Peer Device=2
    Device=1 CAN Access Peer Device=3
    Device=2 CAN Access Peer Device=0
    Device=2 CAN Access Peer Device=1
    Device=2 CAN Access Peer Device=3
    Device=3 CAN Access Peer Device=0
    Device=3 CAN Access Peer Device=1
    Device=3 CAN Access Peer Device=2

    ***NOTE: In case a device doesn't have P2P access to other one,
     it falls back to normal memcopy procedure.
    So you can see lesser Bandwidth (GB/s) in those cases.

    P2P Connectivity Matrix
         D\D     0     1     2     3
         0	     1     1     1     1
         1	     1     1     1     1
         2	     1     1     1     1
         3	     1     1     1     1
    Unidirectional P2P=Disabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 556.45   3.07   3.04   3.04
         1   3.04 552.51   3.04   3.04
         2   3.15   3.15 554.87   3.16
         3   3.15   3.15   3.15 553.29
    Unidirectional P2P=Enabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 555.65   5.18   5.17   5.17
         1   5.18 554.87   5.18   5.17
         2   4.22   4.23 555.65   4.03
         3   4.26   4.26   4.07 554.08
    Bidirectional P2P=Disabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 563.67   5.99   5.87   5.88
         1   6.02 562.46   5.89   5.89
         2   5.91   5.86 561.24   5.98
         3   5.87   5.84   5.98 562.05
    Bidirectional P2P=Enabled Bandwidth Matrix (GB/s)
       D\D     0      1      2      3
         0 565.30   8.21   7.27   7.31
         1   8.21 563.27   7.26   7.30
         2   7.26   7.26 562.86   6.21
         3   7.30   7.31   6.23 562.86
    P2P=Disabled Latency Matrix (us)
       D\D     0      1      2      3
         0   2.97  15.87  16.50  16.44
         1  15.86   2.99  16.42  16.42
         2  17.40  17.18   2.99  18.50
         3  17.31  17.32  18.47   2.96
    P2P=Enabled Latency Matrix (us)
       D\D     0      1      2      3
         0   2.98   5.42   5.90   5.91
         1   5.46   3.02   5.57   5.62
         2   7.03   6.76   3.06   6.66
         3   6.82   6.73   6.65   3.04

附录 B：VGG model train（猫 vs 狗图像集）

    import numpy as np
    np.random.seed(42)

    import keras
    from keras.models import Sequential
    from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D
    from keras.layers.normalization import BatchNormalization
    from keras.callbacks import TensorBoard

    from keras.preprocessing.image import ImageDataGenerator

    from keras.utils import multi_gpu_model

    model = Sequential()

    model.add(Conv2D(64, 3, activation='relu', input_shape=(224,224,3)))
    model.add(Conv2D(64, 3, activation='relu'))
    model.add(MaxPooling2D(2,2))
    model.add(BatchNormalization())

    model.add(Conv2D(128, 3, activation='relu'))
    model.add(Conv2D(128, 3, activation='relu'))
    model.add(MaxPooling2D(2,2))
    model.add(BatchNormalization())

    model.add(Conv2D(256, 3, activation='relu'))
    model.add(Conv2D(256, 3, activation='relu'))
    model.add(Conv2D(256, 3, activation='relu'))
    model.add(MaxPooling2D(2,2))
    model.add(BatchNormalization())

    model.add(Conv2D(512, 3, activation='relu'))
    model.add(Conv2D(512, 3, activation='relu'))
    model.add(Conv2D(512, 3, activation='relu'))
    model.add(MaxPooling2D(2,2))
    model.add(BatchNormalization())

    model.add(Conv2D(512, 3, activation='relu'))
    model.add(Conv2D(512, 3, activation='relu'))
    model.add(Conv2D(512, 3, activation='relu'))
    model.add(MaxPooling2D(2,2))
    model.add(BatchNormalization())

    model.add(Flatten())
    model.add(Dense(4096, activation='tanh'))
    model.add(Dropout(0.5))
    model.add(Dense(4096, activation='tanh'))
    model.add(Dropout(0.5))

    model.add(Dense(1, activation='sigmoid'))

    model.summary()
    _________________________________________________________________
    Layer (type)                 Output Shape              Param #   
    =================================================================
    conv2d_1 (Conv2D)            (None, 222, 222, 64)      1792      
    _________________________________________________________________
    conv2d_2 (Conv2D)            (None, 220, 220, 64)      36928     
    _________________________________________________________________
    max_pooling2d_1 (MaxPooling2 (None, 110, 110, 64)      0         
    _________________________________________________________________
    batch_normalization_1 (Batch (None, 110, 110, 64)      256       
    _________________________________________________________________
    conv2d_3 (Conv2D)            (None, 108, 108, 128)     73856     
    _________________________________________________________________
    conv2d_4 (Conv2D)            (None, 106, 106, 128)     147584    
    _________________________________________________________________
    max_pooling2d_2 (MaxPooling2 (None, 53, 53, 128)       0         
    _________________________________________________________________
    batch_normalization_2 (Batch (None, 53, 53, 128)       512       
    _________________________________________________________________
    conv2d_5 (Conv2D)            (None, 51, 51, 256)       295168    
    _________________________________________________________________
    conv2d_6 (Conv2D)            (None, 49, 49, 256)       590080    
    _________________________________________________________________
    conv2d_7 (Conv2D)            (None, 47, 47, 256)       590080    
    _________________________________________________________________
    max_pooling2d_3 (MaxPooling2 (None, 23, 23, 256)       0         
    _________________________________________________________________
    batch_normalization_3 (Batch (None, 23, 23, 256)       1024      
    _________________________________________________________________
    conv2d_8 (Conv2D)            (None, 21, 21, 512)       1180160   
    _________________________________________________________________
    conv2d_9 (Conv2D)            (None, 19, 19, 512)       2359808   
    _________________________________________________________________
    conv2d_10 (Conv2D)           (None, 17, 17, 512)       2359808   
    _________________________________________________________________
    max_pooling2d_4 (MaxPooling2 (None, 8, 8, 512)         0         
    _________________________________________________________________
    batch_normalization_4 (Batch (None, 8, 8, 512)         2048      
    _________________________________________________________________
    conv2d_11 (Conv2D)           (None, 6, 6, 512)         2359808   
    _________________________________________________________________
    conv2d_12 (Conv2D)           (None, 4, 4, 512)         2359808   
    _________________________________________________________________
    conv2d_13 (Conv2D)           (None, 2, 2, 512)         2359808   
    _________________________________________________________________
    max_pooling2d_5 (MaxPooling2 (None, 1, 1, 512)         0         
    _________________________________________________________________
    batch_normalization_5 (Batch (None, 1, 1, 512)         2048      
    _________________________________________________________________
    flatten_1 (Flatten)          (None, 512)               0         
    _________________________________________________________________
    dense_1 (Dense)              (None, 4096)              2101248   
    _________________________________________________________________
    dropout_1 (Dropout)          (None, 4096)              0         
    _________________________________________________________________
    dense_2 (Dense)              (None, 4096)              16781312  
    _________________________________________________________________
    dropout_2 (Dropout)          (None, 4096)              0         
    _________________________________________________________________
    dense_3 (Dense)              (None, 1)                 4097      
    =================================================================
    Total params: 33,607,233
    Trainable params: 33,604,289
    Non-trainable params: 2,944
    _________________________________________________________________
    #parallel_model = multi_gpu_model(model, gpus=2)
    #parallel_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

**配置数据**

    tensorbrd = TensorBoard('./logs/vgg-1')
    batch_size = 64
    train_dir= './train'
    img_size = (224,224)

    train_datagen = ImageDataGenerator(rescale=1./255)

    train_generator = train_datagen.flow_from_directory(
            train_dir,
            target_size=img_size,
            #batch_size=batch_size,
            # the following batch size loads all 25000 images at once
            batch_size=25000,
            class_mode='binary')
    Found 25000 images belonging to 2 classes.
    for X, Y in train_generator:
        print('data X shape:', X.shape)
        print('labels Y shape:', Y.shape)
        break
    data X shape: (25000, 224, 224, 3)
    labels Y shape: (25000,)
    #parallel_model.fit_generator(train_generator, epochs=4, verbose=1, callbacks=[tensorbrd],
    #                             steps_per_epoch=25000 // batch_size,
    #                             max_queue_size=25000, workers=16)
    #model.fit_generator(train_generator, epochs=4, verbose=1, callbacks=[tensorbrd])
    #parallel_model.fit(X,Y, batch_size=batch_size, epochs=4, verbose=1, callbacks=[tensorbrd])
    model.fit(X,Y, batch_size=batch_size, epochs=4, verbose=1, callbacks=[tensorbrd])

## **相关阅读：**

[\[转载\]GPU 显存与深度学习中批量（batch size）的联系（1080Ti /Titan V/GV100 对比测试）](https://zhuanlan.zhihu.com/p/36212486)

[\[转载\]NVIDIA Titan V 深度学习性能测试（1080Ti/Titan V/Titan V + Tensor cores 对比测试）](https://zhuanlan.zhihu.com/p/35997023)

[\[转载\]NVIDIA Titan V 科学计算和机器学习能力概览（NVIDIA Titan V vs Titan Xp）](https://zhuanlan.zhihu.com/p/32315953)

 <https://zhuanlan.zhihu.com/p/37273066>
